### PR TITLE
Fix del redundant autoreleasepool

### PR DESCRIPTION
**Pull request by dongranGG**
for-Loop does not cause untimely release, these places don't need autoreleasepool

### CODE REVIEW

File: SDWebImage/Core/SDImageCoderHelper.m [hunk 1]
```diff
@@ -141,13 +141,11 @@ + (UIImage *)animatedImageWithFrames:(NSArray<SDImageFrame *> *)frames {
...
-            NSTimeInterval frameDuration = frame.duration;
-            CGImageRef frameImageRef = frame.image.CGImage;
-            NSDictionary *frameProperties = @{(__bridge NSString *)kCGImagePropertyGIFDictionary : @{(__bridge NSString *)kCGImagePropertyGIFDelayTime : @(frameDuration)}};
-            CGImageDestinationAddImage(imageDestination, frameImageRef, (__bridge CFDictionaryRef)frameProperties);
-        }
+        SDImageFrame *frame = frames[i];
+        NSTimeInterval frameDuration = frame.duration;
+        CGImageRef frameImageRef = frame.image.CGImage;
+        NSDictionary *frameProperties = @{(__bridge NSString *)kCGImagePropertyGIFDictionary : @{(__bridge NSString *)kCGImagePropertyGIFDelayTime : @(frameDuration)}};
+        CGImageDestinationAddImage(imageDestination, frameImageRef, (__bridge CFDictionaryRef)frameProperties);
```
Comment (dreampiggy) [inline SDWebImage/Core/SDImageCoderHelper.m:148]: This is OK.

**Comment by dreampiggy** [comment]
I think it's OK to merge with modification to avoid the `prefetcher` changes

The massive `@autoreleasepool` is used in history code to fix the memory peak (which allocate a huge memory in short time).

See #2792

**Review by dongranGG — COMMENTED** [review]
@dreampiggy  `prefetchOperation` is strongly referenced by `self.prefetchQueue`

**Comment by dreampiggy** [comment]
Do you think that the fix in  https://github.com/SDWebImage/SDWebImage/pull/2792 is wrong ?

Any suggestion about the huge memory usage during large prefetch (for example, 100 urls to prefetch)

**Comment by dongranGG** [comment]
Yes,&nbsp;prefetchOperation is strongly referenced by self.prefetchQueue(NSOperationQueue),for-Loop does not cause untimely release,so&nbsp;autoreleasepool is redundant.


And&nbsp;SDAsyncBlockOperation is not a __autoreleasing object,ARC will add release method to release SDAsyncBlockOperation object in for-Loop，if you delete "[self.prefetchQueue addOperation:prefetchOperation]",this autoreleasepool still be redundant


We can test it.


------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "SDWebImage/SDWebImage"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年11月8日(星期二) 晚上6:58
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [SDWebImage/SDWebImage] Fix del redundant autoreleasepool (PR #3388)





 
Do you think that the fix in  #2792 is wrong ?
 
Any suggestion about the huge memory usage during large prefetch (for example, 100 urls to prefetch)

**Comment by dongranGG** [comment]
@dreampiggy 
Yes,`prefetchOperation` is strongly referenced by `self.prefetchQueue`(`NSOperationQueue`),for-Loop does not cause untimely release,so autoreleasepool is redundant.


And `SDAsyncBlockOperation` is not a `__autoreleasing` object,`ARC` will add release method to release `SDAsyncBlockOperation` object in for-Loop，if you delete "`[self.prefetchQueue addOperation:prefetchOperation]`",this autoreleasepool still be redundant.


We can test it.

**Comment by dreampiggy** [comment]
I get your idea about that this autoreleasepool can not solve that stacktrace, but I'm thing the correct way to fix that:

```
Fatal Exception: NSMallocException
0  CoreFoundation                 0x2337b927c __exceptionPreprocess
1  libobjc.A.dylib                0x2329939f8 objc_exception_throw
2  CoreFoundation                 0x2337c6a78 __CFExceptionProem
3  Foundation                     0x234128880 -[NSOperation init]
4  Myapp                          0x102c76c00 -[SDAsyncBlockOperation initWithBlock:] + 25 (SDAsyncBlockOperation.m:25)
5  Myapp                          0x102c76c80 +[SDAsyncBlockOperation blockOperationWithBlock:] + 33 (SDAsyncBlockOperation.m:33)
6  Myapp                          0x102c953bc -[SDWebImagePrefetcher startPrefetchWithToken:] + 112 (SDWebImagePrefetcher.m:112)
7  Myapp                          0x102c951c8 -[SDWebImagePrefetcher prefetchURLs:progress:completed:] + 103 (SDWebImagePrefetcher.m:103)
8  Myapp                          0x1022179b8 -[MYImagePrefetcher addImageUrls:] + 83 (MYImagePrefetcher.m:83)
9  Myapp                          0x1022177cc -[MYImagePrefetcher addFeedModels:resolution:prefetchImgWithWifi:] + 75 (MYImagePrefetcher.m:75)

```

**Comment by dongranGG** [comment]
I don't know why these places add `autoreleasepool`.
I'm just enjoying learn SDWebImage's code,it’s beatful!
And I find some `autoreleasepool` is redundant, so i want fix it like here or elsewhere.
The addition of `autoreleasepool` will cause iOSer to think that there is a delayed release issue.

**Comment by dreampiggy** [comment]
I use Instruments and test 100 urls to prefetch. Seems there are no leak

But if you use the default options, since we will force-decode instead of using lazy-decoding, all the 100 UIImage bitmap buffer need allocate on RAM, this may cause a incresement of RAM.

截屏2022-11-08 21 32 12


@kinarobin Can you provide more information about this ? Or should we, set the default `options` to `SDWebImageAvoidDecodedImage` (or the new context option from SDWebImage 5.14.0, the `SDImageCoderDecodeUseLazyDecoding = YES`) ?

**Comment by dreampiggy** [comment]
After add the `SDWebImageAvoidDecodeImage`, this is the RAM usage, to highest from 147MB -> 23MB.

Since most of UIImage does not render on screen, it's no need to decode just in background in advance.

截屏2022-11-08 21 36 18

**Comment by dreampiggy** [comment]
Just out of topic. I just remove the autoreleasepool in the prefetcher. @dongranGG 

Let we talk about how the large url list should be prefetched in new issue. @kinarobin

**Comment by dongranGG** [comment]
It's cool!👍🏻👍🏻

**Comment by riv1r** [comment]
hi 怎样能够成为Maintainer?😛

**Pull request merged**